tools/cleanup/cleanup.go

```diff
@@ -0,0 +1,308 @@
+			errsLock.RLock()
+			failed := len(errs) > 0
+			errsLock.RUnlock()
+			if failed {
+				return
+			}
+			c, err := deleteFunc(project, before)
+			// Update counter and errors list.
+			countLock.Lock()
+			count += c
+			countLock.Unlock()
+			if err != nil {
+				cause := errors.Cause(err).Error()
+				errsLock.Lock()
+				if _, ok := errs[cause]; !ok {
+					errs[cause] = err
+				}
+				errsLock.Unlock()
+			}
+			<-concurrencyChan
+		}(projects[i])
+	}
+	wg.Wait()
+	// Extract the error strings from the map of errors.
+	errStrings := make([]string, 0)
+	for i := range errs {
+		errStrings = append(errStrings, errs[i].Error())
+	}
+	return count, errStrings
+}
+
+// showStats displays basic stats based on the return values from deleteResources.
+func showStats(count int, errs []string) {
+	log.Printf("%d resources deleted", count)
+	if len(errs) > 0 {
+		log.Printf("%d errors occurred: %s", len(errs), strings.Join(errs, ", "))
+	}
+}
+
+// cleanup performs the main operations and return any errors.
+func cleanup() error {
+	// Sanity check flags
+	if *project == "" && *projectResourceYaml == "" {
+		return errors.New("neither project nor resource file provided")
+	}
+
+	if *project != "" && *projectResourceYaml != "" {
+		return errors.New("provided both project and resource file")
+	}
+
+	if *dryRun {
+		log.Println("-- Running in dry-run mode, no resource deletion --")
+	}
+
+	// Perform authentication if necessary, and create required clients.
+	opts := make([]option.ClientOption, 0)
+	if *serviceAccount != "" {
+		// Activate the service account.
+		if _, output, err := common.ExecCommand("gcloud", "auth", "activate-service-account", "--key-file="+*serviceAccount); err != nil {
+			return fmt.Errorf("cannot activate service account:\n%s", output)
+		}
+		// Create GKE client with specific credentials.
+		opts = append(opts, option.WithCredentialsFile(*serviceAccount))
+	}
+	var err error
+	if gkeClient, err = gke.NewSDKClient(opts...); err != nil {
+		return errors.Wrapf(err, "cannot create GKE SDK client")
+	}
+
+	// Select projects, perform operations.
+	var projects []string
+	if projects, err = selectProjects(*project, *projectResourceYaml, *reProjectName); err != nil {
+		return err
+	}
+	start := time.Now()
+
+	if *daysToKeepImages >= 0 {
+		log.Println("Removing images that are:")
+		log.Printf("- older than %d days", *daysToKeepImages)
+		showStats(deleteResources(projects, start.Add(-24*time.Hour*time.Duration(*daysToKeepImages)), deleteImages))
+	}
+
+	if *hoursToKeepClusters >= 0 {
+		log.Println("Removing clusters that are:")
+		log.Printf("- older than %d hours", *hoursToKeepClusters)
+		showStats(deleteResources(projects, start.Add(-time.Hour*time.Duration(*hoursToKeepClusters)), deleteClusters))
+	}
+
+	log.Printf("All operations finished in %s", time.Now().Sub(start))
+
+	return nil
+}
+
+// main is the script entry point.
+func main() {
+	flag.Parse()
+	if err := cleanup(); err != nil {
```

Comment:
IMHO, to make it easier to implement unit tests, it would be helpful to use something OOP like style, so that the actual functions that need network can be easily mocked, and all functions other than main can have unit tests. something in my mind looks like:
```
imageClient := newImageClient(sa)
imageClient.Run(option. daysToKeepImages...)
...
```

Reply:
Made the suggested change to the approach. Honestly I'm not convinced it makes mocking easier since it's just 2 "classes". PTAL and LMK if that's what you had in mind. The code is untested, but I don't want to invest more time on it before making sure the changes do address your feedback. Once they do, I'll make sure the code works and update the tests.

Reply:
Thanks  for making the change, this is very close to what I had in mind. Will lgtm once the code is tested